pkg/cdi/cdi.go

```diff
@@ -0,0 +1,108 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdi
+
+import (
+	"fmt"
+
+	"github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	cdiSpecs "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	"github.com/golang/glog"
+
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/types"
+)
+
+// CDI represents CDI API required by Device plugin
+type CDI interface {
+	CreateCDISpecForPool(resourcePrefix string, rPool types.ResourcePool) error
+	CreateContainerAnnotations(devicesIDs []string, resourceNamePrefix, resourceKind string) (map[string]string, error)
+}
+
+// Impl implements CDI interface
+type Impl struct {
+}
+
+// CreateCDISpecForPool creates CDI spec file with specified devices
+func (c *Impl) CreateCDISpecForPool(resourcePrefix string, rPool types.ResourcePool) error {
+	cdiDevices := make([]cdiSpecs.Device, 0)
+	cdiSpec := cdiSpecs.Spec{
+		Version: cdiSpecs.CurrentVersion,
+		Kind:    resourcePrefix + "/" + rPool.GetCDIKind(),
+		Devices: cdiDevices,
+	}
+
+	for _, dev := range rPool.GetDevices() {
+		containerEdit := cdiSpecs.ContainerEdits{
+			DeviceNodes: make([]*cdiSpecs.DeviceNode, 0),
+			Env:         make([]string, 0),
+		}
+
+		envs, err := rPool.GetEnvs(resourcePrefix, []string{dev.GetID()})
+		if err != nil {
+			glog.Errorf("CreateCDISpecForPool(): Unable to create environment variables: %s", err)
+			return fmt.Errorf("unable to create environment variables: %s", err)
+		}
+		for name, value := range envs {
+			containerEdit.Env = append(containerEdit.Env, name+"="+value)
+		}
+
+		for _, spec := range rPool.GetDeviceSpecs([]string{dev.GetID()}) {
+			deviceNode := cdiSpecs.DeviceNode{
+				Path:        spec.ContainerPath,
+				HostPath:    spec.HostPath,
+				Permissions: "rw",
+			}
+			containerEdit.DeviceNodes = append(containerEdit.DeviceNodes, &deviceNode)
+		}
+		device := cdiSpecs.Device{
+			Name:           dev.GetID(),
+			ContainerEdits: containerEdit,
+		}
+		cdiSpec.Devices = append(cdiSpec.Devices, device)
+	}
+
+	err := cdi.GetRegistry().SpecDB().WriteSpec(&cdiSpec, resourcePrefix)
```

Comment:
what will happen when we have two resource pools with the same prefix ? will it overwrite ? (i think yes)

Reply:
it will but we shouldn't create two resource pools with the same name